src/storage/src/table/mem_table.rs

```diff
@@ -45,26 +46,66 @@ impl MemTable {
     }
 
     /// read methods
-    pub async fn get_row(&self, pk: &Row) -> StorageResult<Option<RowOp>> {
-        todo!()
+    pub fn get_row(&self, pk: &Row) -> StorageResult<Option<RowOp>> {
+        let res = self.buffer.get(pk);
+        match res {
+            Some(row_op) => Ok(Some(row_op.clone())),
+            None => Ok(None),
+        }
     }
 
     /// write methods
-    pub async fn insert(&mut self, _pk: Row, _value: Row) -> StorageResult<()> {
+    pub fn insert(&mut self, pk: Row, value: Row) -> StorageResult<()> {
+        let entry = self.buffer.entry(pk);
+        match entry {
+            Entry::Vacant(e) => {
+                e.insert(RowOp::Insert(value));
+            }
+            Entry::Occupied(mut e) => match e.get() {
+                RowOp::Delete(old_value) => {
+                    let old_val = old_value.clone();
+                    e.insert(RowOp::Update((old_val, value)));
+                }
+                _ => {
+                    panic!(
+                        "invalid flush status: double insert {:?} -> {:?}",
+                        e.key(),
+                        value
+                    );
+                }
+            },
+        }
         Ok(())
     }
 
-    pub async fn delete(&mut self, _pk: Row, _old_value: Row) -> StorageResult<()> {
+    pub fn delete(&mut self, pk: Row, old_value: Row) -> StorageResult<()> {
+        let entry = self.buffer.entry(pk);
+        match entry {
+            Entry::Vacant(e) => {
+                e.insert(RowOp::Delete(old_value));
+            }
+            Entry::Occupied(mut e) => match e.get() {
+                RowOp::Insert(_) => {
+                    e.remove();
+                }
+                RowOp::Delete(_) => {
+                    panic!(
+                        "invalid flush status: double delete {:?} -> {:?}",
+                        e.key(),
+                        old_value
+                    );
+                }
+                RowOp::Update((old_value, _new_value)) => {
+                    let old_val = old_value.clone();
```

Comment:
why clone?

Reply:
should assert user provided old_value == new_value

Reply:
`old_value` and  `new_value`  are not equal, in this case,  `old_value` is for delete and `new_value` is for insert.

Reply:
user provided old value should always be equal to the original new value.